### PR TITLE
[runtime][python] Add IRPA entry conversion to/from numpy

### PR DESCRIPTION
**Pull request by sogartar**
Add iterop between numpy ndarray and parameter index. This is an adaptation of the original in IREE Turbine https://github.com/iree-org/iree-turbine/blob/142c8a5044a4fedb43a11229f462363b05743b23/iree/turbine/aot/params.py
The goal is to maintain compatibility with IRPA files that were already generated with IREE Turbine.
At some point we can refactor the IREE Turbine side to use this implementation.

### CODE REVIEW

**Review by stellaraccident — APPROVED** [review]
Looks reasonable to me but I'm afk and didn't do a detailed review. Would you move getting a second review (maybe from Scott)?

File: runtime/bindings/python/iree/runtime/io.py
Comment (ScottTodd) [inline runtime/bindings/python/iree/runtime/io.py]: two points here:

* The "nd" in "ndarray" stands for "N-dimensional", not "named": https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html
* grammar nit:  "a named" since "named" does not start with a vowel sound... but then perhaps confusingly "an ndarray" is correct (as shown on that page too), since "ndarray" when spoken _does_ start with a vowel sound "en dee array"

```suggestion
    """Adds an ndarray to the index."""
```

File: runtime/bindings/python/iree/runtime/io.py
Comment (ScottTodd) [inline runtime/bindings/python/iree/runtime/io.py:74]: This file already has `import array`, so this function shadows `array` with a new local variable. 

We have similar code that doesn't also `import array`: https://github.com/iree-org/iree/blob/a31da1f7ab2c81d3fd6eb74f3950d73b56607852/runtime/bindings/python/iree/runtime/system_api.py#L80-L88

I guess this is fine, just be careful :)

File: runtime/bindings/python/iree/runtime/io.py
Comment (ScottTodd) [inline runtime/bindings/python/iree/runtime/io.py]: Can this be changed to `NUMPY:`? This file doesn't depend on torch.

```suggestion
_metadata_prefix = "NUMPY:"
```

You could also move this up in the file, ahead of its first uses

File: runtime/bindings/python/iree/runtime/io.py
Comment (ScottTodd) [inline runtime/bindings/python/iree/runtime/io.py:86]: Oh, I'd nearly forgotten that we have support for arbitrary metadata in parameter archives. So this lets the C runtime load the data as dense bytes without needing to care about the original types, but then Python code can read the metadata to unpack.

On that note, do you want to also write some version information into the metadata, to be able to handle encoding changes more gracefully? Maybe as part of the `_metadata_prefix`, like `NUMPY_v0:`?

**Review by ScottTodd — COMMENTED** [review]
Drive-by since I saw my name. A few minor comments but otherwise LGTM. Thanks!

Please resolve the DCO check: https://github.com/iree-org/iree/pull/19492/checks?check_run_id=34489850995

https://iree.dev/developers/general/contributing/#developer-certificate-of-origin

File: runtime/bindings/python/iree/runtime/io.py
Comment (sogartar) [reply runtime/bindings/python/iree/runtime/io.py]: This is essentially our format magic number. Unfortunately, this can not be change if we want files that were already created with IREE Turbine to be compatible.
At some point we could start writing files with a new prefix, but we would have to maintain support for the old one.
I think we would want to have some versioning scheme as well.

File: runtime/bindings/python/iree/runtime/io.py
Comment (sogartar) [reply runtime/bindings/python/iree/runtime/io.py:86]: I answered on the other question.

File: runtime/bindings/python/iree/runtime/io.py
Comment (sogartar) [reply runtime/bindings/python/iree/runtime/io.py]: I changed the format while maintaining support for loading IREE Turbine tensors.

File: runtime/bindings/python/iree/runtime/io.py
Comment (ScottTodd) [inline runtime/bindings/python/iree/runtime/io.py:200]: (Feel free to merge without this, but it would be nice)

We can probably trust this code... but if you want to go a bit further and guard against compatibility breaking changes, you could add a unit test with some binary data exported from iree-turbine to ensure that it is supported here:

```python
def testParameterIndexEntryFromIreeTurbine(self):
    index = rt.ParameterIndex()
    index.load("test_iree_turbine_index.irpa")  # load from a testdata file
    index_entry_as_array = rt.parameter_index_entry_as_numpy_ndarray(...)
    # then do some sanity check on the returned ndarray
```

(avoiding the temptation to put logic in the test and call the `_make_tensor_metadata` helper function: https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html)

**Review by ScottTodd — APPROVED** [review]

File: runtime/bindings/python/iree/runtime/io.py
Comment (sogartar) [reply runtime/bindings/python/iree/runtime/io.py:200]: I added a test.

Comment (ScottTodd) [reply runtime/bindings/python/iree/runtime/io.py:200]: Nice, thanks!

**Review by ScottTodd — APPROVED** [review]
LGTM once DCO check is resolved (highly recommend setting up SSH commit signature verification on each machine you work from, that shows your commits as verified across all GitHub projects and appeases the DCO check: https://iree.dev/developers/general/contributing/#crypographically-signing-commits)

**Pull request merged**
